File: ext/cpp_polygon_finder/PolygonFinder/src/polygon/finder/Node.cpp
/*
 * Node.cpp
 *
 * Copyright (c) 2025-2026 Emanuele Cesaroni
 *
 * Licensed under the GNU Affero General Public License v3 (AGPLv3).
 * See the LICENSE file in this directory for the full license text.
 */

#include <string.h>
#include <string>
#include <iostream>
#include <algorithm>
#include <limits>
#include <vector>
#include <list>
#include "Node.h"
#include "NodeCluster.h"

Node::Node(int min_x, int max_x, int y, NodeCluster* cluster, char name)
: start_point(min_x, y),
  end_point(max_x, y),
  cluster(cluster) {
  this->name = name;
  this->min_x = min_x;
  this->max_x = max_x;
  this->y = y;
  this->tangs_count = 0;
  this->abs_x_index = 0;
  this->down_indexer = 0;
  this->up_indexer = 0;
  this->track = 0;
  this->outer_index = -1;
  this->inner_index = -1;
}

bool Node::get_trackmax() {
  return((this->track & OMAX) != 0);
}

bool Node::track_complete() {
  return((this->track & OCOMPLETE) == OCOMPLETE);
}

bool Node::track_uncomplete() {
  return((this->track & OCOMPLETE) != OCOMPLETE);
}

void Node::add_intersection(Node& other_node, int other_node_index) {
  if (other_node.y < this->y) {
    upper_start = std::min(upper_start, other_node_index);
    upper_end = std::max(upper_end, other_node_index);
  } else {
    lower_start = std::min(lower_start, other_node_index);
    lower_end = std::max(lower_end, other_node_index);
  }
}

void Node::precalc_tangs_sequences(NodeCluster& cluster) {
  this->tangs_sequence.clear();

  int lower_size = this->lower_end >= 0 ? (this->lower_end - this->lower_start + 1) : 0;
  int upper_size = this->upper_end >= 0 ? (this->upper_end - this->upper_start + 1) : 0;
  this->tangs_sequence.reserve(lower_size + upper_size);

  if (this->upper_end >= 0) {
    this->up_indexer = -cluster.vert_nodes[y + T_UP][this->upper_start].abs_x_index;
  }
  // --- CLOCKWISE (UP) ---
  for (int upper_pos = this->upper_start; upper_pos <= this->upper_end; upper_pos++) {
    tangs_sequence.push_back(-(upper_pos + 1));
  }
  if (this->lower_end >= 0) {
    this->down_indexer = (cluster.vert_nodes[y + T_DOWN][this->lower_start].abs_x_index + lower_size + upper_size - 1);
  }
  // --- COUNTER-CLOCKWISE (DOWN) ---
  for (int lower_pos = this->lower_end; lower_pos >= this->lower_start; lower_pos--) {
    tangs_sequence.push_back(lower_pos);
  }
  this->tangs_count = this->tangs_sequence.size();
}

Node* Node::my_next_inner(Node *last, int versus) {
  unsigned int last_node_index;
  if (last->y < this->y)       last_node_index = last->abs_x_index + this->up_indexer;
  else            last_node_index = this->down_indexer - last->abs_x_index;
  if (versus == Node::O) last_node_index == 0 ? last_node_index = this->tangs_sequence.size() - 1 : last_node_index--;
  else          last_node_index == this->tangs_sequence.size() - 1 ? last_node_index = 0 : last_node_index++;
  return get_tangent_node_by_virtual_index(this->tangs_sequence[last_node_index]);
}

Node* Node::my_next_outer(Node *last, int versus) {
  unsigned int last_node_index;
  if (last->y < this->y) last_node_index = last->abs_x_index + this->up_indexer;
  else                   last_node_index = this->down_indexer - last->abs_x_index;
  if (versus == Node::O) last_node_index == this->tangs_sequence.size() - 1 ? last_node_index = 0 : last_node_index++;
  else                   last_node_index == 0 ? last_node_index = this->tangs_sequence.size() - 1 : last_node_index--;
  return get_tangent_node_by_virtual_index(this->tangs_sequence[last_node_index]);
}

Point* Node::coords_entering_to(Node *enter_to, int mode, int tracking) {
  int enter_to_index;
  if (enter_to->y < this->y) enter_to_index = enter_to->abs_x_index + this->up_indexer;
  else                       enter_to_index = this->down_indexer - enter_to->abs_x_index;

  int tg_index = this->tangs_sequence[enter_to_index];
  Point* point;
  if (tg_index < 0) {
    Node& node_up = cluster->vert_nodes[y + T_UP][-(tg_index + 1)];
    if (mode == Node::A) {
      enter_to->track |= TURNER[tracking][OMAX - 1];
      point = &node_up.end_point;
    } else {
      enter_to->track |= TURNER[tracking][OMIN - 1];
      point = &node_up.start_point;
    }
  } else {
    Node& node_down = cluster->vert_nodes[y + T_DOWN][tg_index];
    if (mode == Node::A) {
      enter_to->track |= TURNER[tracking][OMIN - 1];
      point = &node_down.start_point;
    } else {
      enter_to->track |= TURNER[tracking][OMAX - 1];
      point = &node_down.end_point;
    }
  }
  return point;
}

Node* Node::get_tangent_node_by_virtual_index(int virtual_index) {
  if (virtual_index < 0) {
    return &(this->cluster->vert_nodes[y + T_UP][-(virtual_index + 1)]);
  } else {
    return &(this->cluster->vert_nodes[y + T_DOWN][virtual_index]);
  }
}

File: lib/contrek/finder/node.rb
module Contrek
  module Finder
    class Node
      include Listable

      attr_reader :min_x, :max_x, :y, :name, :tangs_sequence, :tangs_count, :data_pointer,
        :upper_start, :upper_end, :lower_start, :lower_end, :start_point, :end_point
      attr_accessor :track, :abs_x_index, :outer_index, :inner_index

      T_UP = -1
      T_DOWN = 1
      OMAX = 1 << 0
      OMIN = 1 << 1
      IMAX = 1 << 2
      IMIN = 1 << 3
      OCOMPLETE = OMIN | OMAX

      TURN_MAX = IMAX | OMAX
      TURN_MIN = IMIN | OMIN

      # 0 = outer
      # 1 = inner
      TURNER = [[OMAX, OMIN], [TURN_MAX, TURN_MIN]]

      OUTER = 0
      INNER = 1

      def initialize(cluster, min_x, max_x, y, name, connectivity_offset = 0)
        @name = name
        @min_x = min_x
        @max_x = max_x
        @y = y
        @tangs_sequence = nil
        @tangs_count = 0
        @track = 0
        @abs_x_index = 0
        @data_pointer = cluster.lists.get_data_pointer
        @up_indexer = 0
        @down_indexer = 0
        @outer_index = -1
        @inner_index = -1
        @upper_start = Float::INFINITY
        @upper_end = -1
        @lower_start = Float::INFINITY
        @lower_end = -1
        @start_point = {x: min_x, y: y}
        @end_point = {x: max_x, y: y}
        @cluster = cluster
        cluster.add_node(self, connectivity_offset)
      end

      def get_tangent_node_by_virtual_index(virtual_index)
        return nil if virtual_index.nil?
        if virtual_index < 0
          @cluster.vert_nodes[y + T_UP][-(virtual_index + 1)]
        else
          @cluster.vert_nodes[y + T_DOWN][virtual_index]
        end
      end

      def my_next(last, versus, mode)
        last_node_index = if last.y < y
          last.abs_x_index + @up_indexer
        else
          @down_indexer - last.abs_x_index
        end

        case mode
        when :outer
          if versus == :o
            (last_node_index == tangs_sequence.size - 1) ? last_node_index = 0 : last_node_index += 1
          else
            (last_node_index == 0) ? last_node_index = tangs_sequence.size - 1 : last_node_index -= 1
          end

        when :inner
          if versus == :o
            (last_node_index == 0) ? last_node_index = tangs_sequence.size - 1 : last_node_index -= 1
          else
            (last_node_index == tangs_sequence.size - 1) ? last_node_index = 0 : last_node_index += 1
          end
        end
        get_tangent_node_by_virtual_index(@tangs_sequence.at(last_node_index))
      end

      def coords_entering_to(enter_to, enter_mode, tracking)
        enter_to_index = if enter_to.y < y
          enter_to.abs_x_index + @up_indexer
        else
          @down_indexer - enter_to.abs_x_index
        end

        tg_index = @tangs_sequence[enter_to_index]
        if tg_index < 0
          node_up = @cluster.vert_nodes[y + T_UP][-(tg_index + 1)]
          if enter_mode == :a
            enter_to.track |= TURNER[tracking][OMAX - 1]
            point = node_up.end_point
          else
            enter_to.track |= TURNER[tracking][OMIN - 1]
            point = node_up.start_point
          end
        else
          node_down = @cluster.vert_nodes[y + T_DOWN][tg_index]
          if enter_mode == :a
            enter_to.track |= TURNER[tracking][OMIN - 1]
            point = node_down.start_point
          else
            enter_to.track |= TURNER[tracking][OMAX - 1]
            point = node_down.end_point
          end
        end
        point
      end

      def track_uncomplete
        (@track & OCOMPLETE) != OCOMPLETE
      end

      def track_complete
        (@track & OCOMPLETE) == OCOMPLETE
      end

      def add_intersection(other_node, other_node_index)
        if other_node.y < y
          @upper_start = other_node_index if other_node_index < @upper_start
          @upper_end = other_node_index if other_node_index > @upper_end
        else
          @lower_start = other_node_index if other_node_index < @lower_start
          @lower_end = other_node_index if other_node_index > @lower_end
        end
      end

      def precalc_tangs_sequences(cluster:)
        @tangs_sequence = []
        @up_indexer = -cluster.vert_nodes[@y + T_UP][@upper_start].abs_x_index if @upper_end >= 0
        if @upper_end >= 0
          (@upper_start..@upper_end).each do |upper_pos|
            @tangs_sequence << -(upper_pos + 1)
          end
        end
        if @lower_end >= 0
          lower_size = (@lower_end >= 0) ? (@lower_end - @lower_start + 1) : 0
          upper_size = (@upper_end >= 0) ? (@upper_end - @upper_start + 1) : 0
          @down_indexer = (cluster.vert_nodes[@y + T_DOWN][@lower_start].abs_x_index + lower_size + upper_size - 1)
        end
        if @lower_end >= 0
          @lower_end.downto(@lower_start).each do |lower_pos|
            @tangs_sequence << lower_pos
          end
        end
        @tangs_count = tangs_sequence.size
      end
    end
  end
end
